src/sidebar/config/build-settings.js

```diff
@@ -0,0 +1,147 @@
+import { getApiUrl } from './get-api-url';
+import { hostPageConfig } from './host-config';
+import * as postMessageJsonRpc from '../util/postmessage-json-rpc';
+
+/**
+ * @typedef {import('../../types/config').ConfigFromSidebar} ConfigFromSidebar
+ * @typedef {import('../../types/config').ConfigFromHost} ConfigFromHost
+ * @typedef {import('../../types/config').SidebarSettings} SidebarSettings
+ */
+
+/**
+ * Returns the global embedder ancestor frame.
+ *
+ * @param {number} levels - Number of ancestors levels to ascend.
+ * @param {Window=} window_
+ * @return {Window}
+ */
+function getAncestorFrame(levels, window_ = window) {
+  let ancestorWindow = window_;
+  for (let i = 0; i < levels; i++) {
+    if (ancestorWindow === ancestorWindow.top) {
+      throw new Error(
+        'The target parent frame has exceeded the ancestor tree. Try reducing the `requestConfigFromFrame.ancestorLevel` value in the `hypothesisConfig`'
+      );
+    }
+    ancestorWindow = ancestorWindow.parent;
+  }
+  return ancestorWindow;
+}
+
+/**
+ * Retrieve host configuration and list of groups from `targetFrame` via RPC
+ *
+ * @param {Window} targetFrame
+ * @param {string} origin - Origin filter for `window.postMessage` call
+ * @return {Promise<ConfigFromHost>}
+ */
+async function fetchConfigRPC(targetFrame, origin) {
+  const configFromHost = await postMessageJsonRpc.call(
+    targetFrame,
+    origin,
+    'requestConfig',
+    [],
+    3000
+  );
+  // Closure for the RPC call to scope parentFrame and origin variables.
+  const rpcCall = (method, args = [], timeout = 3000) =>
+    postMessageJsonRpc.call(targetFrame, origin, method, args, timeout);
+  return fetchGroupsAsync(configFromHost, rpcCall);
+}
+
+/**
+ * Check the host configuration object to see if the list of groups should be
+ * retrieved via RPC. If so, replace the `groups` value with the results of
+ * an applicable RPC call.
+ *
+ * This allows the app to start with an incomplete config and then lazily
+ * fill in the `groups` value(s) later when its ready. This helps speed
+ * up the loading process.
+ *
+ * @param {ConfigFromHost} configFromHost - Host configuration object. This
+ *   may be mutated with a list of groups fetched by RPC.
+ * @param {function} rpcCall - RPC method
+ *  (method, args, timeout) => Promise
+ * @return {Promise<ConfigFromHost>} - The mutated host configuration
+ */
+async function fetchGroupsAsync(configFromHost, rpcCall) {
+  if (Array.isArray(configFromHost.services)) {
+    configFromHost.services.forEach((service, index) => {
+      if (service.groups === '$rpc:requestGroups') {
+        // Groups should be fetched by RPC. Do not set a timeout as we may
+        // be waiting on user input (e.g. re-authorization).
+        service.groups = rpcCall('requestGroups', [index], null).catch(() => {
+          throw new Error('Unable to fetch groups');
+        });
+      }
+    });
+  }
+  return configFromHost;
+}
+
+/**
```

Comment:
Everything here and below can be considered "new construction": rework of the previous `fetchConfig` function.